### ISSUE TITLE
`kdb mount`: Using Absolute Filepath Does Not Work Correctly When Mounting  to `user` Namespace

### ISSUE THREAD

**Issue by sanssecours** [ElektraInitiative/libelektra#975]
#  Step to Reproduce

1. Enter the following command inside a shell

```sh
kdb mount /tmp/user.ini user/test
```

# Expected Behaviour

Elektra now store every key below the `user/test` namespace in `/tmp/user.ini`. 

# Actual Behaviour

Elektra stores the keys below `user/test` in `$HOME/tmp/user.ini`. We can verify this problem via

```
kdb file user/test
```

which returns `/Users/rene/tmp/user.ini` on my machine. What makes this worse, is that `kdb mount` returns the wrong path:

```
/tmp/user.ini on user/test with name user/test
```

I found this usability problem when I discovered that running `testshell_mathcheck` creates  a directory that looks something like this:  `/Users/rene/var/folders/hx/flbncdhj4fs87095gzxvnj3h0000gn/T` on my machine.

## The Good

It looks like mounting to `system` behaves correctly as the following terminal session shows (`»` marks the command prompt):

```sh
» kdb mount /tmp/system.ini system/test
» kdb mount
/tmp/system.ini on system/test with name system/test
» kdb file system/test
/tmp/system.ini
```

## The Bad

If we continue after the terminal session above, then the behaviour looks pretty bad to me:

```
» kdb mount /tmp/user.ini user/test
» kdb mount
/tmp/system.ini on system/test with name system/test
/tmp/user.ini on user/test with name user/test
» kdb file user/test
/Users/rene/tmp/user.ini
» kdb file system/test
/tmp/system.ini
```

## The Ugly

I wasted many hours debugging this, and would like my time back.

# System Information

- Elektra Version: master
- OS: macOS 10.12

**Comment by markus2330** [comment]
Thank you for the detailed analysis and the detailed description of your expectations.

> I wasted many hours debugging this, and would like my time back.

Please report issues earlier. Unfortunately we did not have time last meeting to discuss the namespaces because you obviously have a different understanding of what they should be. Shortly put they are chroots of where the configuration stems from (or more generally: which semantics they have).

Obviously the current docu does not describe it well enough, because you read most parts and still had a different expectation. What about enforcing `~` at the beginning of absolute user pathes (and disallowing absolute pathes)? What do you expect for `dir` pathes?

Btw. did you also read "kdb info resolver"?

**Comment by sanssecours** [comment]
> Unfortunately we did not have time last meeting to discuss the namespaces because you obviously have a different understanding of what they should be. 

If you think the behaviour described above is fine, then this is certainly the case 😀.

> What about enforcing `~` at the beginning of absolute user paths (and disallowing absolute pathes)? 

I think we should just use the same behaviour of (nearly) every other unix utility out there: A path that start with `/` is an absolute path. This would also add the benefit that nobody would have to write/read a text about this special behaviour. At least I would have expected a warning about this. Something like:

```markdown
You possibly wanted to store data in `/tmp/user.ini`. I would have done that if you
specified the namespace `system`. But since you used the namespace `user` I will
just save you data in `~/tmp/user.ini` instead. Muahaha 😈.

TL;DR: Elektra is much smarter than you!
```

> Btw. did you also read "kdb info resolver"?

Nope, but I read a lot of the other documentation, and still I wasted many hours debugging this stuff. I think I am not alone with my expectations.

@KurtMi: Did you expect that `mathcheck.esr` creates temporary directories – of a form similar to `var/folders/hx/flbncdhj4fs87095gzxvnj3h0000gn/T` – inside your home directory?

> What do you expect for `dir` pathes?

I never used the `dir` namespace. For the time being, I expect that I will not use this feature at all 😀.

**Comment by markus2330** [comment]
> I think we should just use the same behaviour of (nearly) every other unix utility out there: A path that start with / is an absolute path. This would also add the benefit that nobody would have to write/read a text about this special behaviour. At least I would have expected a warning about this.

I agree that the usability is bad here and we need a change in the `kdb mount` tool. That paths starting with / are absolute in UNIX is only true to a certain degree: Chroot was invented about 1979 and gave a possibility to make absolute paths relative. The purpose of namespaces is similar: If you want to mount something outside your home (user) or current (dir) directory, you need to use the system-namespace.

What was your expectations of the differences of the namespaces? Why did you want to use `user`-namespace if the mountpoint had nothing to do with your user?

> I think I am not alone with my expectations.

Certainly not, but we have to think about a way how to fulfil these expectations without breaking expectations for cascading (or other) mountpoints.

> I never used the dir namespace. 

For a consistent user experience we need a behaviour which works for all namespaces.

**Comment by sanssecours** [comment]
> That paths starting with / are absolute in UNIX is only true to a certain degree: Chroot was invented about 1979 and gave a possibility to make absolute paths relative. 

You learn something new every day. Thanks for the information.

> What was your expectations of the differences of the namespaces? 

I did not expect the namespaces to come into play at all when using an absolute path.

> Why did you want to use user-namespace if the mountpoint had nothing to do with your user?

I just used the `user` namespace here, since this is the same namespace that `mathcheck.esr` specifies.

**Comment by KurtMi** [comment]
> @KurtMi: Did you expect that mathcheck.esr creates temporary directories – of a form similar to var/folders/hx/flbncdhj4fs87095gzxvnj3h0000gn/T – inside your home directory?

The temp file will be created with:

    TMPFILE=$(mktemp -t elektraenv.XXXXXXXXX 2>/dev/null || mktemp -t 'elektraenv')

On my system (Arch) the file goes to:

    /tmp/elektraenv.TjKv4VfZO

It seems weird that the file ends up in your home dir, but honestly I am not sure how osx treats this command.

**Comment by markus2330** [comment]
As I understood @sanssecours the problem is not what `mktemp` does, but what Elektra does when "this" file is mounted to `user`-namespace.

**Comment by markus2330** [comment]
I tried to improve it in ded6db01a7d48829b1a6e6b1dbaa9fb753983125

Is this sufficient?

**Comment by sanssecours** [comment]
> Is this sufficient?

Looks good to me. Thank you.
